### PR TITLE
HDR Output support

### PR DESCRIPTION
**Pull request by FrancescoC-unity**
_NOTE:_ PR has lots of files that have lots of actual changes compressed by github, please when reviewing expand the diff. 

This PR introduces proper HDR Output support for HDRP. 

TL;DR of the features and characteristics: 

- Support for HDR10 and scRGB output standard. 
- ACES workflow for HDR (3 presets, 1000, 2000 and 4000 nits)
- Adding Full ACES even for LDR workflow.
- Neutral tonemapper using standard BT2390 curve for range reduction.
- Added control over how much hue shift is desired. 
- Control over target min/max nits and display paper white. 
- Proper UI handling so that white in UI is maintained as white.
- Added several debug visualizations, with special focus on gamut visualization. 
- Changed the whole grading pipeline to always use ACEScg as working space and allowing to go back to sRGB. 
- Optimized LUT building so it is created only when a change happens. 

When enabling HDR all the HDR Output tunable parameters are shown in the Tonemapping volume component
image
image

Unfortunately cannot post images as they wouldn't show much on LDR screens and LDR screenshots :D The debug view can be seen here though

image
image
image

Most of the math for conversion and tone curves are in [HDROutput.hlsl ](https://github.com/Unity-Technologies/Graphics/blob/a3a7d7fb76b1914922104e6f887a626a0a366ae5/com.unity.render-pipelines.core/ShaderLibrary/HDROutput.hlsl) I tried to be verbose with comments there and added a bunch of desmos 

A lot more can be written, but I am keeping this PR description tight. Please reach out and I am happy to discuss any aspect in more details. 

I will be start writing docs now and will probably open a separate doc PR when I am done. Though I also need to find a way to fake a LDR/HDR comparison in LDR for Docs, and I have no clue how to do that :D

### CODE REVIEW

**Comment by FrancescoC-unity** [comment]
Given the change in color grading I am expecting a few failed tests, I am running yamato and fix the failing ones.

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/HDRenderPipeline.PostProcess.cs [hunk 1]
```diff
@@ -284,12 +289,14 @@ void CleanupPostProcess()
             CoreUtils.Destroy(m_ClearBlackMaterial);
             CoreUtils.Destroy(m_SMAAMaterial);
             CoreUtils.Destroy(m_TemporalAAMaterial);
+            CoreUtils.Destroy(m_GradingAndTonemappingLUT);
```
Comment (JulienIgnace-Unity) [inline com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/HDRenderPipeline.PostProcess.cs:292]: Should be RTHandles.Release here (or m_GradingAndTonemappingLUT.Release)

**Review by JulienIgnace-Unity — APPROVED** [review]

**Comment by FrancescoC-unity** [comment]
> Please update What's new and upgrade guide (or do it in the doc PR if it is not done there :) ) (in particular for this: Changed the whole grading pipeline to always use ACEScg as working space and allowing to go back to sRGB.)

Added it in the doc PR: https://github.com/Unity-Technologies/Graphics/pull/5992/files#diff-e64740d8eca588a80a48a41a51c9f66e6b56062814a5b273b4b399cae0a5a689R3

File: com.unity.render-pipelines.core/ShaderLibrary/ACES.hlsl [hunk 2]
```diff
@@ -501,16 +507,19 @@ half segmented_spline_c5_fwd(half x)
...
+    float coefsLow[10];    // coefs for B-spline between minPoint and midPoint (units of log luminance)
+    float coefsHigh[10];   // coefs for B-spline between midPoint and maxPoint (units of log luminance)
+    half2 minPoint; // {luminance, luminance} linear extension below this
+    half2 midPoint; // {luminance, luminance}
+    half2 maxPoint; // {luminance, luminance} linear extension above this
+    float slopeLow;       // log-log slope of low linear extension
+    float slopeHigh;      // log-log slope of high linear extension
+};
 
+half segmented_spline_c9_fwd(half x, SegmentedSplineParams_c9 params)
```
Comment (sebastienlagarde) [inline com.unity.render-pipelines.core/ShaderLibrary/ACES.hlsl:521]: Note: we will need to mention this change in upgrade guide

File: com.unity.render-pipelines.core/ShaderLibrary/ACES.hlsl [hunk 3]
```diff
@@ -762,14 +771,27 @@ static const half ODT_SAT_FACTOR = 0.93;
 //
 half3 ODT_RGBmonitor_100nits_dim(half3 oces)
 {
+    const SegmentedSplineParams_c9 ODT_48nits =
+    {
+        // coefsLow[10]
+        { -1.6989700043, -1.6989700043, -1.4779000000, -1.2291000000, -0.8648000000, -0.4480000000, 0.0051800000, 0.4511080334, 0.9113744414, 0.9113744414},
```
[the remark this answers is not in the document]
Comment (sebastienlagarde) [reply com.unity.render-pipelines.core/ShaderLibrary/ACES.hlsl:777]: also if I am not wrong this code is duplicated few time, can't we declare it a single time?

[the remark this answers is not in the document]
Comment (FrancescoC-unity) [reply com.unity.render-pipelines.core/ShaderLibrary/ACES.hlsl:777]: Originally I had it statically defined, but it generated obscene code, I can have a function returning the factors though, will do.

File: com.unity.render-pipelines.core/ShaderLibrary/HDROutput.hlsl [hunk 4]
```diff
@@ -0,0 +1,657 @@
...
+    float3 lms = RotateRec2020ToLMS(Rec2020);
+    float3 PQLMS = LinearToPQ(max(0.0f, lms));
+    return RotatePQLMSToICtCp(PQLMS);
+}
+
+
+
+float3 RotateOutputSpaceToICtCp(float3 inputColor)
+{
+    // TODO: Do the conversion directly from Rec709 (bake matrix Rec709 -> XYZ -> LMS)
```
Comment (sebastienlagarde) [inline com.unity.render-pipelines.core/ShaderLibrary/HDROutput.hlsl:197]: guess the compiler may do it?

File: com.unity.render-pipelines.core/ShaderLibrary/HDROutput.hlsl [hunk 5]
```diff
@@ -0,0 +1,657 @@
...
+//  OETFs
+// --------------------------------
+// The functions here are OETF, technically for applying the opposite of the PQ curve, we are mapping
+// from linear to PQ space as this is what the display expects.
+// See this desmos for comparisons https://www.desmos.com/calculator/5jdfc4pgtk
+#define PRECISE_PQ 0
+#define ISS_APPROX_PQ 1
+#define GTS_APPROX_PQ 2
+
+#define OETF_CHOICE GTS_APPROX_PQ
```
Comment (sebastienlagarde) [inline com.unity.render-pipelines.core/ShaderLibrary/HDROutput.hlsl:303]: is it worse it? the PRECISE_PQ dont look like so expensive. Or maybe change depends on platform? approx on console but full on PC?

File: com.unity.render-pipelines.core/ShaderLibrary/HDROutput.hlsl (hunk 4 above)
Comment (FrancescoC-unity) [reply com.unity.render-pipelines.core/ShaderLibrary/HDROutput.hlsl:197]: This is a bit different, what I mean is essentially have a version that is RotateRec2020ToICtCp without needing the extra step in between. 

I am not sure the compiler will be able to merge matrices by itself. 

That said, this is not a big deal as it all happen during LUT building.

File: com.unity.render-pipelines.core/ShaderLibrary/HDROutput.hlsl (hunk 5 above)
Comment (FrancescoC-unity) [reply com.unity.render-pipelines.core/ShaderLibrary/HDROutput.hlsl:303]: It is still a few trascendentals, the approx is very very similar (see the desmos https://www.desmos.com/calculator/5jdfc4pgtk )

File: com.unity.render-pipelines.core/ShaderLibrary/HDROutput.hlsl [hunk 6]
```diff
@@ -0,0 +1,657 @@
...
+        hueShiftedResult.y = BT2390EETF(input.y, minNits, maxNits);
+        hueShiftedResult.z = BT2390EETF(input.z, minNits, maxNits);
+    }
+    return hueShiftedResult;
+}
+
+// Ref "High Dynamic Range
color grading and display
in Frostbite" [Fry 2017]
+float3 FryHuePreserving(float3 input, float minNits, float maxNits, float hueShift, int mode)
```
Comment (sebastienlagarde) [inline com.unity.render-pipelines.core/ShaderLibrary/HDROutput.hlsl:485]: not used anywhere?

File: com.unity.render-pipelines.core/ShaderLibrary/HDROutput.hlsl (hunk 5 above)
Comment (FrancescoC-unity) [reply com.unity.render-pipelines.core/ShaderLibrary/HDROutput.hlsl:303]: Check the error diff in the diff, it is very very minimal (bottom of the desmos)

File: com.unity.render-pipelines.core/ShaderLibrary/HDROutput.hlsl (hunk 6 above)
Comment (FrancescoC-unity) [reply com.unity.render-pipelines.core/ShaderLibrary/HDROutput.hlsl:485]: No, but I'd like to keep as I expect some feedback at some point about this hue preservation. I rolled our own for the same thing, but if we need to revisit I'd like to have this here.

File: com.unity.render-pipelines.high-definition/Editor/PostProcessing/TonemappingEditor.cs [hunk 7]
```diff
@@ -125,6 +152,57 @@ public override void OnInspectorGUI()
...
+                PropertyField(m_UseFullACES);
+            }
+
+            if (hdrInPlayerSettings)
+            {
+                EditorGUILayout.LabelField("HDR Output");
+                int hdrTonemapMode = m_Mode.value.intValue;
+                if (m_Mode.value.intValue == (int)TonemappingMode.Custom || hdrTonemapMode == (int)TonemappingMode.External)
+                {
+                    EditorGUILayout.HelpBox("The selected tonmapping mode is not supported in HDR Output mode. Select a fallback mode.", MessageType.Warning);
```
Comment (sebastienlagarde) [inline com.unity.render-pipelines.high-definition/Editor/PostProcessing/TonemappingEditor.cs:166]: typo tonmapping

File: com.unity.render-pipelines.high-definition/Runtime/Debug/DebugHDR.shader [hunk 8]
```diff
@@ -0,0 +1,333 @@
...
+
+        float p = (a + b + c) * 0.5;
+        float h = 2 / c * sqrt(p * (p - a) * (p - b) * (p - c));
+
+
+        float lineAlpha =  lerp(1.0, 0.0, smoothstep(0.5 * thickness, 1.5 * thickness, h));
+        return float4(color * lineAlpha, lineAlpha);
+    }
+
+    float3 xyYtoXYZ(float3 xyY)
```
Comment (sebastienlagarde) [inline com.unity.render-pipelines.high-definition/Runtime/Debug/DebugHDR.shader:98]: don't we have those function in color.hlsl? we should move them there

File: com.unity.render-pipelines.high-definition/Runtime/PostProcessing/Components/Tonemapping.cs [hunk 9]
```diff
@@ -20,7 +20,7 @@ public enum TonemappingMode
         Neutral,
 
         /// <summary>
-        /// Close approximation of the reference ACES tonemapper for a more filmic look.
+        /// ACES tonemapper for a more filmic look.
```
Comment (sebastienlagarde) [inline com.unity.render-pipelines.high-definition/Runtime/PostProcessing/Components/Tonemapping.cs:23]: to clarify, it mean we don't have approx anymore for ACES? (guess you told me but I forget), does it mean we have a perf hit here?

Comment (FrancescoC-unity) [reply com.unity.render-pipelines.high-definition/Runtime/PostProcessing/Components/Tonemapping.cs:23]: We have both now, approx or not depends on a checkbox (scroll a bit down the file for the checkbox)

Comment (FrancescoC-unity) [reply com.unity.render-pipelines.high-definition/Runtime/PostProcessing/Components/Tonemapping.cs:23]: By default the checkbox is off, no need for anything to do when upgrading. Isn't it more a what's new?

**Review by sebastienlagarde — APPROVED** [review]

File: com.unity.render-pipelines.high-definition/Runtime/Debug/DebugHDR.shader (hunk 8 above)
Comment (sebastienlagarde) [reply com.unity.render-pipelines.high-definition/Runtime/Debug/DebugHDR.shader]: ok

File: com.unity.render-pipelines.core/ShaderLibrary/ACES.hlsl (hunk 2 above)
Comment (FrancescoC-unity) [reply com.unity.render-pipelines.core/ShaderLibrary/ACES.hlsl:521]: Yeah to be safe I added the old signature/version so it is compatible and should work like before

File: com.unity.render-pipelines.high-definition/Runtime/PostProcessing/Shaders/DebugHDRxyMapping.compute [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+#define GROUP_SIZE_Y 8
+
+TEXTURE2D_X(_SourceTexture);
+RW_TEXTURE2D_X(float, _xyBuffer);
+
+float4 _HDRxyBufferDebugParams;
+#define _SizePerDim _HDRxyBufferDebugParams.xy
+#define _IsRec709 (int)(_HDRxyBufferDebugParams.w == 1)
+
+float2 RGBtoxy(float3 rgb)
```
[the remark this answers is not in the document]
Comment (FrancescoC-unity) [reply com.unity.render-pipelines.high-definition/Runtime/PostProcessing/Shaders/DebugHDRxyMapping.compute:18]: Moved at least partially

**Pull request merged**